text/0000-bee-pow/0000-bee-pow.md

```diff
@@ -0,0 +1,332 @@
+a valid nonce, so that my transaction is accepted and propagated by all honest nodes in the network.
+
+When I - as a node - receive a network message from the gossip pipeline, I want to be able to check if its contained
+nonce value satisfies the pre-defined network difficulty, so that I can decide whether to discard or process that
+message.
+
+# Detailed design
+
+## Hasher
+A hash function always operates on some kind of input data, usually but not necessarily on some kind of `u8` array - and deterministically produces another piece of data - the hash - for example a `u64` that usually takes much less space, and therefore can be used as a label for the data it was derived from. In Rust such a hash function can be generalized using the following trait:
+
+```Rust
+pub trait Hasher {
+    type Data;
+    type Hash;
+
+    fn hash(&mut self, data: &Self::Data) -> Self::Hash;
+}
+```
+
+Notive that by making `Data` and `Hash` associated types this RFC doesn't make any commitment to the underlying representation which allows for a very flexible design.
+
+## Sampler
+
+The `Sampler` trait allows implementing different types that mutate the input data handed to the hasher on each try. The standard sampler used in IOTA simply starts with a zero nonce and increments it until it finds a valid one. There are, however, other samplers possible, like a random sampler.
+
+```Rust
+pub trait Sampler {
+    type Data;
+
+    fn next(&mut self, data: &mut Self::Data);
+}
+```
+
+Like in the `Hasher` trait we now have to make `Data` an associated type as well, so it can always be compatible when used together in `PearlDiver`.
+
+## Tester
+
+The `Tester` trait allows implementing different types that place  plug different validation schemes into `PearlDiver`, like the `Hashcash` or the `Hamming` validation scheme.
+
+```Rust
+pub trait Tester {
+    type Hash;
+
+    fn is_valid(&self, hash: &Self::Hash) -> bool;
+}
+```
+
+As with `Sampler` the `Tester` trait needs an associated type so it can always be implemented in a compatible way and operate on something served by a type implementing the `Hasher` trait. 
+
+## PearlDiver
+
+With the traits above a type can now be defined that is called `PearlDiver`. It encapsulates the following algorithm:
+
+Starting with some serialized transaction data (which also encodes a zero nonce):
+1. hash the data
+2. check if the hash satisifies some pre-defined and adjustable constraint
+   1. hash satisfies constraint => stop the algorithm 
+   2. hash doesn't satisfy constraint 
+      1. select another nonce 
+      2. goto 1.
+
+This RFC proposes the following `PearlDiver` type:
+
+```Rust
+pub struct PearlDiver<H, S, T, D, R>
+where
+    H: Hasher<Data = D, Hash = R>,
+    S: Sampler<Data = D>,
+    T: Tester<Hash = R>,
+{
+    hasher: H,
+    sampler: S,
+    tester: T,
+}
+
+impl<H, S, T, D, R> PearlDiver<H, S, T, D, R>
```

Comment:
As long as I pretty love these trait layers, I feel like generics in `PearlDiver` is a little bit out of hand. Would it be better to use trait object in this case? Don't get me wrong, I actually prefer static dispatch over dynamic. Just wanna bring some suggestions and see if it fits in. I would like to hear others' opinion between trait bound and trait objects.

```suggestion
struct PearlDiver<D, R> {
    hasher: Box<dyn Hasher<Data = D, Hash = R>>,
    sampler: Box<dyn Sampler<Data = D>>,
    tester: Box<dyn Tester<Hash = R>>,
}
```